ssl/ech/ech_internal.c

```diff
@@ -160,4 +216,1107 @@ int ossl_ech_conn_init(SSL_CONNECTION *s, SSL_CTX *ctx,
+    length_with_snipadding = innersnipadding + encoded_len;
+    length_of_padding = 31 - ((length_with_snipadding - 1) % 32);
+    length_with_padding = encoded_len + length_of_padding + innersnipadding;
+    /*
+     * Finally - make sure final result is longer than padding target
+     * and a multiple of our padding increment.
+     * TODO(ECH): This is a local addition - we might take it out if
+     * it makes us stick out; or if we take out the above more (uselessly:-)
+     * complicated scheme, we may only need this in the end.
+     */
+    if (length_with_padding % OSSL_ECH_PADDING_INCREMENT)
+        length_with_padding += OSSL_ECH_PADDING_INCREMENT
+            - (length_with_padding % OSSL_ECH_PADDING_INCREMENT);
+    while (length_with_padding < OSSL_ECH_PADDING_TARGET)
+        length_with_padding += OSSL_ECH_PADDING_INCREMENT;
+    clear_len = length_with_padding;
+    OSSL_TRACE_BEGIN(TLS) {
+        BIO_printf(trc_out, "EAAE: padding: mnl: %zu, lws: %d "
+                   "lop: %d, lwp: %d, clear_len: %zu, orig: %zu\n",
+                   mnl, length_with_snipadding, length_of_padding,
+                   length_with_padding, clear_len,
+                   encoded_len);
+    } OSSL_TRACE_END(TLS);
+    return clear_len;
+}
+
+/*
+ * Calculate AAD and do ECH encryption
+ * pkt is the packet to send
+ * return 1 for success, other otherwise
+ *
+ * 1. Make up the AAD: the encoded outer, with ECH ciphertext octets zero'd
+ * 2. Do the encryption
+ * 3. Put the ECH back into the encoding
+ * 4. Encode the outer (again!)
+ */
+int ossl_ech_aad_and_encrypt(SSL_CONNECTION *s, WPACKET *pkt,
+                             unsigned char *encoded_inner,
+                             size_t encoded_inner_len)
+{
+    int rv = 0, hpke_mode = OSSL_HPKE_MODE_BASE;
+    OSSL_ECHSTORE_ENTRY *ee = NULL;
+    OSSL_HPKE_SUITE hpke_suite = OSSL_HPKE_SUITE_DEFAULT;
+    unsigned char config_id_to_use = 0x00, info[SSL3_RT_MAX_PLAIN_LENGTH];
+    unsigned char *clear = NULL, *cipher = NULL, *aad = NULL, *mypub = NULL;
+    size_t cipherlen = 0, aad_len = 0, lenclen = 0, mypub_len = 0;
+    size_t info_len = SSL3_RT_MAX_PLAIN_LENGTH, clear_len = 0;
+
+    if (s == NULL)
+        return 0;
+    if (s->ext.ech.es == NULL || s->ext.ech.es->entries == NULL
+        || pkt == NULL || s->ssl.ctx == NULL) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    rv = ossl_ech_pick_matching_cfg(s, &ee, &hpke_suite);
+    if (rv != 1 || ee == NULL) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    s->ext.ech.attempted_type = ee->version;
+    OSSL_TRACE_BEGIN(TLS) {
+        BIO_printf(trc_out, "EAAE: selected: version: %4x, config %2x\n",
+                   ee->version, ee->config_id);
+    } OSSL_TRACE_END(TLS);
+    config_id_to_use = ee->config_id; /* if requested, use a random config_id instead */
+    if ((s->ssl.ctx->options & SSL_OP_ECH_IGNORE_CID) != 0
```

Comment:
Why are you checking both the ctx and ssl options? We would normally only check the ssl options

Reply:
why checking? most likely out of ignorance:-) If only checking `s->options` is more correct, no problem here so I'll do that.  That reminds me that I don't have tests of `SSL_OP_ECH_IGNORE_CID` in the test harness (and it's been ages since I tested that), so I added a `TODO(ECH)`:` to `test/ech_test.c` to remind me to add something there, meanwhile i'll change to just checking `s->options`.